### DISCUSSION TITLE
How to solve reportMissingTypeStubs for custom modules ?

### DISCUSSION THREAD

**Discussion by guillaumecherel** [microsoft/pyright#3168] [Q&A] [answered]
Hi, 

In strict mode, pyright reports missing type stubs for my custom modules. For example :
```  
/home/guillaume-pro/Work/oplc/api.py:3:6 - error: Stub file not found for "oplc.constants" (reportMissingTypeStubs)
```
I can't figure out how to deal with this. I don't think I'm supposed to create stubs for my own files when I put type annotations in the source, am I ?

**Comment by erictraut** [comment]
What do you mean by "custom modules"? Are these modules part of your local project, or are they installed in your Python environment (e.g. in `site-packages`)?

If these modules are part of your local project, are they written in Python, or are they binary files compiled from some other language like C++?

What is the root of your project (i.e. the folder you opened in your editor)? Is it `Work` or `oplc`?

If you add `"verboseOutput": true` to your pyrightconfig.json, it will emit additional log information when it fails to resolve an import. You might find this helpful in tracking down the issue.

**Comment by guillaumecherel** [comment, marked as answer]
Thanks for your help !

By local modules, I mean modules of my local project.

I realized that when I run pyright from the root of my python project, those errors don't appear. But I run pyright from a higher-level directory, my root project that contains the python project and others. My project tree looks like:
```
- myproject/
- myproject/frontend/...: (javascript project)
- myproject/backend/ (the python project)
- myproject/backend/mypackage/ 
- myproject/backend/mypackage/mymodule.py
- myproject/backend/mypackage/... (a total of 8 modules)

```
When I run pyright from `myproject/backend/`, it works fine, but I'd like to run it from `myproject/`. (This is where I run neovim with the coc-pyright plugin, so I suppose neovim runs pyright from there.)

I tried to configure pyright to run it from `myproject/` by adding a configuration file `myproject/pyrightconfig.json` with the following:
```
{
    "include": ["backend/mypackage"]
    "exclude": [
        "**/__pycache__"
    ],
    "typeCheckingMode": "strict"
}
```
 
I get errors like
```
/home/guillaume/Work/myproject/backend/mypackage/mymodule.py:3:6 - error: Stub file not found for "mypackage.mymodule" (reportMissingTypeStubs)
```

**Reply by erictraut** [reply]
You need to tell pyright about the additional search paths. By default, it includes only the root of your project. You can do this by adding your subdirectories to the "extraPaths" key in the pyrightconfig.json file. For details, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#main-pyright-config-options) and [this documentation](https://github.com/microsoft/pyright/blob/main/docs/import-resolution.md).

**Reply by guillaumecherel** [reply]
Thanks ! That solved it.

I had tried the extraPath but hadn't used it right. What worked was to point it to the root of the python project:

```
...
   "extraPaths": ["backend/"]
...
```
